File: src/Worker/content/NBB.Worker.FSharp/DependencyInjectionExtensions.fs
[<AutoOpen>]
module ServiceCollectionExtensions

open Microsoft.Extensions.Configuration
open Microsoft.Extensions.DependencyInjection
open NBB.Worker.Application
open NBB.Messaging.Host
open NBB.Messaging.Host.Builder
open NBB.Messaging.Host.MessagingPipeline
open NBB.Messaging.Abstractions
open NBB.Application.Mediator.FSharp
open NBB.Core.Effects
open System
open NBB.Worker.Messaging
#if NatsMessagingTransport
open NBB.Messaging.Nats
open System.Reflection
open Microsoft.Extensions.Logging
#if OpenTracing
open NBB.Messaging.OpenTracing.Subscriber
open NBB.Messaging.OpenTracing.Publisher
open OpenTracing
open OpenTracing.Noop
open Jaeger
open Jaeger.Samplers
open Jaeger.Reporters
open Jaeger.Senders.Thrift
open OpenTracing.Util
#endif
#endif

type IServiceCollection with
    member this.AddWorkerServices(configuration: IConfiguration) =

        // Application
        this |> WriteApplication.addServices |> ignore

        // Messaging
#if NatsMessagingTransport
        this
            .AddMessageBus()
            .AddNatsTransport(configuration)
        |> ignore
#endif

        this.Decorate<IMessageBusPublisher, SamplePublisherDecorator>() |> ignore

#if OpenTracing
        this.Decorate<IMessageBusPublisher, OpenTracingPublisherDecorator>()
        |> ignore
#endif
        this.AddMessagingHost
            (fun hostBuilder ->
                hostBuilder.Configure
                    (fun configBuilder ->
                        configBuilder
                            .AddSubscriberServices(fun config ->
                                config.AddTypes(typeof<HelloWorld.Command>)
                                |> ignore)
                            .WithDefaultOptions()
                            .UsePipeline(fun pipelineBuilder ->
                                pipelineBuilder
                                    .UseCorrelationMiddleware()
                                    .UseExceptionHandlingMiddleware()
#if OpenTracing
                                    .UseMiddleware<OpenTracingMiddleware>()
#endif
#if Resiliency
                                    .UseDefaultResiliencyMiddleware()
#endif
                                    .UseMiddleware<SampleSubscriberPipelineMiddleware>()
                                    .UseEffectMiddleware(fun m ->
                                        m
                                        |> Mediator.sendMessage
                                        |> EffectExtensions.ToUnit)
                                |> ignore)
                        |> ignore)
                |> ignore)
        |> ignore


#if OpenTracing
        // OpenTracing
        this.AddOpenTracing() |> ignore

        this.AddSingleton<ITracer>(fun (serviceProvider : IServiceProvider) ->
            let isEnabled = configuration.GetValue<bool>("OpenTracing:Jeager:IsEnabled")
            if not isEnabled then
                NoopTracerFactory.Create();
            else
                let serviceName = Assembly.GetEntryAssembly().GetName().Name
                let loggerFactory = serviceProvider.GetRequiredService<ILoggerFactory>()
                let tracer = 
                    Tracer.Builder(serviceName)
                        .WithLoggerFactory(loggerFactory)
                        .WithSampler(ConstSampler true)
                        .WithReporter(
                            RemoteReporter.Builder()
                                .WithSender(
                                    UdpSender(
                                        configuration.GetValue<string>("OpenTracing:Jeager:AgentHost"),
                                        configuration.GetValue<int>("OpenTracing:Jeager:AgentPort"), 0))
                                .Build())
                        .Build() :> ITracer

                GlobalTracer.Register(tracer) |> ignore

                tracer
        )
#endif

File: src/Worker/templates/NBB.Worker.CSharp/DependencyInjectionExtensions.cs
﻿using MediatR;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using NBB.Worker.Messaging;
using NBB.Worker.Application;
using NBB.Messaging.Host;
using NBB.Messaging.Abstractions;
#if NatsMessagingTransport
using NBB.Messaging.Nats;
#endif
#if RusiMessagingTransport
#endif
using System.Reflection;
using Microsoft.Extensions.Logging;
#if OpenTracing
using NBB.Messaging.OpenTracing.Subscriber;
using NBB.Messaging.OpenTracing.Publisher;
using OpenTracing;
using OpenTracing.Noop;
using Jaeger;
using Jaeger.Samplers;
using Jaeger.Reporters;
using Jaeger.Senders.Thrift;
using OpenTracing.Util;
#endif


namespace NBB.Worker
{
    public static class DependencyInjectionExtensions
    {
        public static void AddWorkerServices(this IServiceCollection services, IConfiguration configuration)
        {
            // MediatR 
            services.AddMediatR(typeof(HelloWorld.Command).Assembly);
            services.Scan(scan => scan.FromAssemblyOf<HelloWorld.Handler>()
                .AddClasses(classes => classes.AssignableTo(typeof(IPipelineBehavior<,>)))
                .AsImplementedInterfaces()
                .WithScopedLifetime());

            // Messaging
#if NatsMessagingTransport
            services.AddMessageBus().AddNatsTransport(configuration);
#endif
#if RusiMessagingTransport
            services.AddMessageBus().AddRusiTransport(configuration);
#endif            

            services.Decorate<IMessageBusPublisher, SamplePublisherDecorator>();

#if OpenTracing
            services.Decorate<IMessageBusPublisher, OpenTracingPublisherDecorator>();
#endif
            services.AddMessagingHost(
                configuration, 
                hostBuilder => hostBuilder
                .Configure(configBuilder => configBuilder
                    .AddSubscriberServices(selector =>
                    {
                        selector
                            .FromMediatRHandledCommands()
                            .AddAllClasses();
                    })
                    .WithDefaultOptions()
                    .UsePipeline(pipelineBuilder => pipelineBuilder
                        .UseCorrelationMiddleware()
                        .UseExceptionHandlingMiddleware()
#if OpenTracing
                        .UseMiddleware<OpenTracingMiddleware>()
#endif
#if Resiliency
                        .UseDefaultResiliencyMiddleware()
#endif
                        .UseMediatRMiddleware()
                        .UseMiddleware<SampleSubscriberPipelineMiddleware>())));

#if OpenTracing
            // OpenTracing
            services.AddOpenTracing();

            services.AddSingleton<ITracer>(serviceProvider =>
            {
                if (!configuration.GetValue<bool>("OpenTracing:Jeager:IsEnabled"))
                {
                    return NoopTracerFactory.Create();
                }

                string serviceName = Assembly.GetEntryAssembly().GetName().Name;

                ILoggerFactory loggerFactory = serviceProvider.GetRequiredService<ILoggerFactory>();

                ITracer tracer = new Tracer.Builder(serviceName)
                    .WithLoggerFactory(loggerFactory)
                    .WithSampler(new ConstSampler(true))
                    .WithReporter(new RemoteReporter.Builder()
                        .WithSender(new UdpSender(
                            configuration.GetValue<string>("OpenTracing:Jeager:AgentHost"),
                            configuration.GetValue<int>("OpenTracing:Jeager:AgentPort"), 0))
                        .Build())
                    .Build();

                GlobalTracer.Register(tracer);

                return tracer;
            });
#endif
        }
    }
}
